### DISCUSSION TITLE
Portenta H7 TCP Latency Issue

### DISCUSSION THREAD

**Discussion by yannick-t** [micropython/micropython#13518] [STM32 / Pyboard] [answered]
Hello,

I ran into a weird problem with TCP Connections on the Portenta H7 and I couldn't find any mention of it online
TCP Connections seem to have a lot of latency (up to ~500ms).
I ran into it using the requests module, and experimented a bit. UDP connections have a RTT that I would expect, around 1ms over Ethernet.

Here are some measurements that should make it clear
```python
>>> measure_rtt()
Received b'Hello, server'
Round-trip time: 511.019 ms
>>> measure_rtt()
Received b'Hello, server'
Round-trip time: 306.276 ms
>>> measure_rtt()
Received b'Hello, server'
Round-trip time: 408.367 ms
>>> measure_rtt()
Received b'Hello, server'
Round-trip time: 508.853 ms
>>> measure_rtt()
Received b'Hello, server'
Round-trip time: 512.186 ms
>>> measure_rtt()
Received b'Hello, server'
Round-trip time: 73.175 ms
>>> measure_rtt()
Received b'Hello, server'
Round-trip time: 510.502 ms
>>> measure_rtt()
Received b'Hello, server'
Round-trip time: 512.188 ms
>>> 
```

The code I used:
Client (on the Portenta):
```python
def measure_rtt(host='address', port=21212):
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((host, port))
    start_time = time.ticks_us()
    s.sendall(b'Hello, server')
    data = s.recv(13)
    end_time = time.ticks_us()
    print('Received', repr(data))
    print('Round-trip time:', time.ticks_diff(end_time, start_time) / 1000, 'ms')
    s.close()
```
Server (on a PC):
```python
def start_server(host='0.0.0.0', port=21212):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind((host, port))
        s.listen()
        while True:
            conn, addr = s.accept()
            with conn:
                print('Connected by', addr)
                while True:
                    data = conn.recv(1024)
                    if not data:
                        break
                    conn.sendall(data)
```

Has anyone else run into this problem? 
The round trip time cutoff seems to be awfully close to **512ms**, to me it seems like a polling interval somewhere

**Comment by IHOXOHI** [comment]
Hi,
I have just bought the portenta H7 board... If I succeed to install the firmware I could try with mine... But until now, I have an error when I upload the firmware...See the errors on the att
![portentaH7](https://github.com/micropython/micropython/assets/63340434/0a4ae64b-04c7-43c4-b198-afbc3b8adf69)
ached files. If someone can help me...
The error for the portenta H7 is "No rules to make target"...
Other errors with the portenta C33...?:
![portentaC33](https://github.com/micropython/micropython/assets/63340434/f5f783b3-3823-4b7b-b76f-623388e04c5c)
![portentaC332](https://github.com/micropython/micropython/assets/63340434/728dc402-2f9d-44e8-a07c-b36107cfc722)

Thanks.

**Comment by robert-hh** [comment]
@IHOXOHI Did you run `git submodule init;git submodule update` in the stm32 directory? I can build the firmware fine here.
To enable dfu download you have to double tap the blue button on the board. You can check if the board is in download mode with `dfu-util -l`.

**Comment by IHOXOHI** [comment]
Hi Robert,
I'm honored to see you again...
No, I had tried only the Make submodules command... So I will... I presume the result....
Again thanks a lots.

**Comment by IHOXOHI** [comment]
During the time that I have your attention,
I will be so enthusiast if I can use rshell on an android phone... Could you help me for doing this too?
Now, it seems that we have only the virtual possibility...:(

**Reply by robert-hh** [reply]
I know almost nothing about using Smart Phone for something else than working with the built-in apps like Phone, browser, WhatsApp. I played a while ago with a Linux emulation on the smart phone, but that did not get me an access to the USB port as serial device, as far as I recall. Same with an iPad. A simple Laptop or chromebook may be better suited for that usage.

**Reply by glenn20** [reply]
> I will be so enthusiast if I can use rshell on an android phone... Could you help me for doing this too?
Now, it seems that we have only the virtual possibility...:(

You may be interested in the [Micro REPL](https://github.com/orgs/micropython/discussions/11249) app.

**Reply by IHOXOHI** [reply]
For sure. Thanks.

**Comment by IHOXOHI** [comment]
INFO: this build requires mboot to be installed first

**Reply by IHOXOHI** [reply]
During very little time, I have used the appi "UserLAnd" with the archlinux distri, ..., there was an acces to the dev port!!! But, just a litlle moment... Now it seems to be a "milicious" app...
It could be a  good project for micropython development.

**Reply by robert-hh** [reply]
> INFO: this build requires mboot to be installed first

The config files for Arduino Portenta H7 tell, that mboot is not used.

b.t.w.: I used Termius for a while. Looked at UserLand as well, but did not like it.

**Comment by IHOXOHI** [comment]
The logs...
![portentaH72](https://github.com/micropython/micropython/assets/63340434/9047954a-451d-48ff-a0da-979c9b673591)
Thanks.

**Reply by robert-hh** [reply]
Maybe I look at the wrong file. Yes, it expects mboot or another bootloader. The instruction on the firmware download files do not mention it. The board config files do not mention MBOOT, but the stm32h747.ld file mentions an Arduino bootloader. So I assume it's present in the device when sold. Do you get in dfu mode when you push the blue button at power-on?
@iabdalkader  Any help?

**Reply by iabdalkader** [reply]
Arduino boards don't use mboot. Did you switch the board to DFU mode ? Double tap the reset button the green LED will blink, then `make BOARD=.... deploy`

**Comment by IHOXOHI** [comment]
Yes, I think. The led change of state... The ACM0 is present.

**Reply by robert-hh** [reply]
In dfu mode /dev/ttyACM0 is NOT present.

**Comment by IHOXOHI** [comment]
So the board isn't.
But the led pass in other state... Strange.....

**Comment by IHOXOHI** [comment]
I have pressed twice the reset button, the led has changed of color and rythme, and the ACM0 is still present.

**Reply by robert-hh** [reply]
what is the output of
dfu-util -l

**Comment by IHOXOHI** [comment]
The first one is after double press the reset button.
The second, before.
THANKS.
![Uploading 2024-02-27-100850_1920x1080_scrot.png…]()

**Reply by robert-hh** [reply]
Could you try running `dfu-util -l` with sudo after pushing reset twice?

**Comment by IHOXOHI** [comment]
![2024-02-27-104647_1920x1080_scrot](https://github.com/micropython/micropython/assets/63340434/ce7cab2f-bf0e-49b2-96fb-a4c11290318a)

**Reply by robert-hh** [reply]
So the device is in dfu mode. You should be able to download the firmware, maybe try again using `sudo dfu-util ......`

**Comment by IHOXOHI** [comment]
YES!
It works like this...
Again, a sincere Thanks for your percious help.
Have a fun!

**Comment by IHOXOHI** [comment]
The board is impressive!
Maybe this board could do her own repl session...

**Comment by IHOXOHI** [comment]
So,
I tested a TCP connection for the first time with micropython...
I used a raspberry connected to a hat explorer by a RJ45 connection...
server code:
"
import socket
HOST = "127.0.0.1"  # Standard loopback interface address (localhost)
PORT = 23  # Port to listen on (non-privileged ports are > 1023)

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.bind((HOST, PORT))
    s.listen()
    conn, addr = s.accept()
    with conn:
        print(f"Connected by {addr}")
        while True:
            data = conn.recv(1024)
            if not data:
                break
            conn.sendall(data)
"
client:
"
import socket
import time
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(('127.0.0.1', 23))
start_time = time.ticks_us()
s.sendall(b'Hello, server')
data = s.recv(13)
end_time = time.ticks_us()
print('Received', repr(data))
print('Round-trip time:', time.ticks_diff(end_time, start_time) / 1000, 'ms')
s.close()
 "
I tried too with others port numbers...
Always the same result for the client:
"
MicroPython v1.22.1 on 2024-01-05; Arduino Portenta H7 with STM32H747
Type "help()" for more information.
>>> import client3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "client3.py", line 5, in <module>
OSError: [Errno 113] EHOSTUNREACH 
"
Maybe I have made a mistake for the connection...?
Thanks.

**Comment by IHOXOHI** [comment]
Hey,
I just tried to use 2 i2c connections on the hat portenta carrier... One no works. There is 3 address (8,54,80) without anything connected... I think several peripherals... But just my oled no works...  
Is there a way for stop these peripherals. This is on the line 1 (I2C(1))? Without the carrier, the board has 2 address without peripheral (8,54).
Thanks.

**Comment by IHOXOHI** [comment]
Yo,
For the TCP connection, I can see a cursor on the hat explorer with the name E1H CENTER TAP...? Maybe, I need to push it on the on postion? No? I will try again... I don't find documentation on this purpose...

**Comment by dpgeorge** [comment, marked as answer]
Getting back on topic here, I can reproduce the original reported latency issue of around 500ms.  It occurs on any MicroPython port that uses lwIP (which includes the Portenta H7).

The reason is because of this lwIP code in `src/include/lwip/priv/tcp_priv.h`:
```c
#ifndef TCP_TMR_INTERVAL
#define TCP_TMR_INTERVAL       250  /* The TCP timer interval in milliseconds. */
#endif /* TCP_TMR_INTERVAL */

#ifndef TCP_FAST_INTERVAL
#define TCP_FAST_INTERVAL      TCP_TMR_INTERVAL /* the fine grained timeout in milliseconds */
#endif /* TCP_FAST_INTERVAL */

#ifndef TCP_SLOW_INTERVAL
#define TCP_SLOW_INTERVAL      (2*TCP_TMR_INTERVAL)  /* the coarse grained timeout in milliseconds */
#endif /* TCP_SLOW_INTERVAL */
```
Data that's written to the TCP socket is not actually committed to the wire (sent as an ETH packet) until `TCP_SLOW_INTERVAL` time has elapsed, which is 500ms by default.

Sending lots of data through a TCP socket has relatively high bandwidth, because this 500ms timeout is only used for small, on-off writes of data.

You can try rebuilding the Portenta firmware with this interval changed.  In `ports/stm2/lwip_inc/lwipopts.h` add a line:
```c
#define TCP_TMR_INTERVAL 50
```

**Reply by yannick-t** [reply]
Hi thank you very much for looking into this, we ended up using udp for everything in our project but this is good to know in the future.

Also maybe a bit off-topic, I found that with non-blocking udp socket.recvfrom it seems like only one message is buffered independently of size, so calling it multiple times until EAGAIN comes up only ever receives one message. Everything else is discarded. Do you know if there is also some option that could cause this?

**Reply by dpgeorge** [reply]
Commit b1e960270213a88eb02219d85201198c4a8671c6 has basically fixed this TCP latency issue, by enabling the Nagle algorithm.

You can now also do `socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)` to ensure there is no delay between writing to a TCP socket and that data being sent out over the network.

> I found that with non-blocking udp socket.recvfrom it seems like only one message is buffered independently of size

Right... currently the lwIP bindings in MicroPython only store one outstanding incoming UDP packet.  That could be improved to have a queue of packets.

**Comment by IHOXOHI** [comment]
Hey,
I have found the doc for the switch off the tcp on the hat...
On other hand, I see the rgb led like just an on/off... The intensity parameter seems strange on the H7... On the giga too... the same pyb lib... I tried to use directly a pwm command... No pwm class??? Ther is a Timer class but, no works like expected... Is there documentaion...
If someone has an advise...
Thanks.

**Reply by Josverl** [reply]
May I suggest starting a seperate discussion on that topic rather than piling on this one?
